src/GafferScene/Instancer.cpp

```diff
@@ -2651,69 +2696,46 @@ void Instancer::InstancerCapsule::render( IECoreScenePreview::Renderer *renderer
-			// we can reuse the context. But timeOffset is relative, so it's important that we reset the
-			// time before we do setPrototypeContextVariables for the next element. ( Should this be more
-			// general instead of assuming that frame is the only variable for which offsetMode may be set? )
-			prototypeScope.setFrame( onFrameTime );
 
-			engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
-			auto inserted = prototypeContextsMap.emplace( prototypeScope.context()->hash(), prototypeContextsMap.size() );
-
-			if( inserted.second )
-			{
-				prototypes.push_back(
-					addPrototypeToVectors(
-						prototypesPlug, prototypeRoot, prototypeScope, sampleTimes, capsuleHash, *renderOptions
-					)
-				);
-			}
+	const IECore::MurmurHash outerCapsuleHash = [this](){
+		IECore::MurmurHash result;
+		hash( result );
+		return result;
+	}();
 
-			prototypeIndices.push_back( inserted.first->second );
+	const bool hasAttributes = engines[0]->numInstanceAttributes() > 0;
 
-		}
-	}
-	else
+	// If constantPrototype is set, then every instance will use the same prototype.
+	ConstPrototypePtr constantPrototype;
+	if( !engines[0]->hasContextVariables() )
 	{
-		prototypes.push_back(
-			addPrototypeToVectors(
-				prototypesPlug, prototypeRoot, scope, sampleTimes, capsuleHash, *renderOptions
-			)
+		constantPrototype = new Prototype(
+			prototypesPlug, prototypeRoot, sampleTimes, outerCapsuleHash, renderOptions,
+			scope.context(), renderer,
+			// If we don't have instance attributes, we can prepare renderer attributes ahead of time
+			!hasAttributes
 		);
 	}
 
-	const bool hasAttributes = engines[0]->numInstanceAttributes() > 0;
-
-	// If we don't have per-instance attributes, we can pre-prepare renderer attributes ready to go that
-	// depend only on the prototype
-	std::vector< IECoreScenePreview::Renderer::AttributesInterfacePtr > prototypeRendererAttribs;
-	if( !hasAttributes )
-	{
-		for( const Prototype &p : prototypes )
+	// If constantPrototype is not set, we will put prototypes in this cache whenever we first encounter
+	// a prototype using a given context.
+	IECorePreview::LRUCache<IECore::MurmurHash, ConstPrototypePtr, IECorePreview::LRUCachePolicy::Parallel, PrototypeCacheGetterKey> prototypeCache(
+		[
+			&prototypesPlug, &prototypeRoot, &sampleTimes, &outerCapsuleHash, &renderOptions,
+			&renderer, &hasAttributes
+		]
+		( const PrototypeCacheGetterKey &key, size_t &cost, const IECore::Canceller *canceller ) -> ConstPrototypePtr
 		{
-			prototypeRendererAttribs.push_back( renderer->attributes( p.attributes.get() ) );
-		}
-	}
+			cost = 1;
+			return new Prototype(
+				prototypesPlug, prototypeRoot, sampleTimes, outerCapsuleHash, renderOptions,
+				key.context, renderer,
+				// If we don't have instance attributes, we can prepare renderer attributes ahead of time
+				!hasAttributes
+			);
+		},
+		pointIndicesForPrototype.size() // Never evict, even if prototypes are all unique
```

Comment:
Another minor thing, but I'd be tempted to use `limits<size_t>::max()` here so I could move the initialisation of `pointIndicesForPrototype` down into the "output the instances" section. I cope better when stuff lives in smaller scopes.

Reply:
Addressed.